angular-language-server: init at 18.2.0

tricktron opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Packages the angular language server as binary because packaging from source is especially difficult (see https://github.com/NixOS/nixpkgs/pull/349521 for more details).

Closes https://github.com/NixOS/nixpkgs/issues/244019 and closes https://github.com/NixOS/nixpkgs/pull/349521.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bornedj committed:
angular-language-server: init at 18.2.0

SuperSandro2000 commented on pkgs/by-name/an/angular-language-server-bin/package.nix:
```diff
@@ -0,0 +1,83 @@
+    description = "LSP for angular completions, AOT diagnostic, quick info and go to definitions";
+    homepage = "https://github.com/angular/vscode-ng-language-service";
+    changelog = "https://github.com/angular/vscode-ng-language-service/blob/${finalAttrs.version}/CHANGELOG.md";
+    license = lib.licenses.mit;
```
Not sure if we have an appropriate sourceProvenance for prebuild js things.

tricktron replied:
@SuperSandro2000 I don't know either. I put MIT because the nodejs package lists it as MIT as well: https://www.npmjs.com/package/@angular/language-server.

SuperSandro2000 commented on pkgs/by-name/an/angular-language-server-bin/package.nix:
```diff
@@ -0,0 +1,83 @@
+      ];
+      text = ''
+        LATEST_VERSION=$(curl -H "Accept: application/vnd.github+json" \
+          -Ls https://api.github.com/repos/${owner}/${repo}/releases/latest | \
```
This should optionally take a GITHUB_TOKEN

tricktron replied:
@SuperSandro2000 Do you have an example on how to achieve this?

SuperSandro2000 commented on pkgs/by-name/an/angular-language-server-bin/package.nix:
```diff
@@ -0,0 +1,83 @@
+  passthru = {
+    tests = {
+      start-ok = runCommand "${finalAttrs.pname}-test" { } ''
+        ${angular-language-server-bin}/bin/ngserver --stdio --help &> $out
```
We can use lib.getExe here and grep in one go

tricktron replied:
Added lib.getExe.

SuperSandro2000 commented on pkgs/by-name/an/angular-language-server-bin/package.nix:
```diff
@@ -0,0 +1,83 @@
+  installPhase = ''
+    runHook preInstall
+    install -Dm755 server/bin/ngserver $out/bin/ngserver
+    install -Dm755 server/index.js $out/index.js
```
Can we place those not at /?

tricktron replied:
@SuperSandro2000 No because the upstream unwrapped `ngserver.bin` calls the index.js file with the relative path `..`. What would be the advantage of having them both at $out?

SuperSandro2000 replied:
I meant can we place them somewhere else like $out/share/angular-langugage-server so that don't pollute $out

tricktron replied:
Alright, I put them under $out/lib.

SuperSandro2000 commented on pkgs/by-name/an/angular-language-server-bin/package.nix:
```diff
@@ -0,0 +1,83 @@
+}:
+
+let
+  owner = "angular";
```
Just hardcore those. They will likely never ever change.

tricktron replied:
You are right that they probably never change but I think it does not hurt either so I keep them in if that is ok.

SuperSandro2000 reviewed (commented):
We are not arch Linux, please drop the bin suffix.

tricktron commented:
> We are not arch Linux, please drop the bin suffix.

I have put this because there are already some packages with the -bin suffix in nixpkgs, e.g. vlc-bin, vault-bin and all the prebuilt jdks-bins. Do we have another naming convention to differentiate binary builds from source builds for the same package?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4727

SuperSandro2000 commented:
> I have put this because there are already some packages with the -bin suffix in nixpkgs, e.g. vlc-bin, vault-bin and all the prebuilt jdks-bins.

If there is no source variant for those, the suffix there should also go.

> Do we have another naming convention to differentiate binary builds from source builds for the same package?

yes, set meta.sourceProvenance. We don't name pacakages differently because we usually do not want both variants even. Why have a binary package if there is a source one? There are only very little exceptions.

tricktron commented:
> > I have put this because there are already some packages with the -bin suffix in nixpkgs, e.g. vlc-bin, vault-bin and all the prebuilt jdks-bins.
> 
> If there is no source variant for those, the suffix there should also go.
> 
> > Do we have another naming convention to differentiate binary builds from source builds for the same package?
> 
> yes, set meta.sourceProvenance. We don't name pacakages differently because we usually do not want both variants even. Why have a binary package if there is a source one? There are only very little exceptions.

@SuperSandro2000 I dropped the -bin suffix and added the meta.sourceProvenance attribute. Thank you, I learnt something new 👍 .

tricktron commented:
IMHO this is ready to be merged. @lelgenio @NotAShelf do you have time to build this on your machines for a review?

lelgenio reviewed (approved):
Working fine with helix:
![image](https://github.com/user-attachments/assets/535cca47-eeaa-43c8-b887-be136d5f0bad)

lelgenio commented:
Result of `nixpkgs-review pr 350257` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>angular-language-server</li>
  </ul>
</details>

NotAShelf reviewed (approved):
Builds and runs fine

lelgenio commented:
Result of `nixpkgs-review pr 350257` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>angular-language-server</li>
  </ul>
</details>

tricktron commented:
I think this is ready to merge. 

Pinging @wegank and @SuperSandro2000 with merge rights. 

Btw: Is there a better general approach to ping people with merge rights? E.g. is there a nixos group with merge rights?

wegank commented:
##### Reviewed points

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on aarch64-darwin
- [x] executables tested on aarch64-darwin
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [ ] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] `meta.mainProgram` is set, if applicable.
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [ ] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

##### Possible improvements

Set `meta.platforms` to `lib.platforms.unix` to drop the `10.rebuild-darwin: 0` label.

##### Comments

LGTM except the point above.

tricktron commented:
> Set meta.platforms to lib.platforms.unix to drop the 10.rebuild-darwin: 0 label.

@wegank Done.

SuperSandro2000 commented:
#354195
